modules/pubxaiAnalyticsAdapter.js

```diff
@@ -13,196 +23,335 @@ const defaultHost = 'api.pbxai.com';
+    dealId: bidResponse.dealId,
+    currency: bidResponse.currency,
+    floorData: bidResponse.floorData,
+    mediaType: bidResponse.mediaType,
+    netRevenue: bidResponse.netRevenue,
+    requestTimestamp: bidResponse.requestTimestamp,
+    responseTimestamp: bidResponse.responseTimestamp,
+    status: bidResponse.status,
+    sizes: parseSizesInput(bidResponse.size).toString(),
+    statusMessage: bidResponse.statusMessage,
+    timeToRespond: bidResponse.timeToRespond,
+    transactionId: bidResponse.transactionId,
+    bidId: bidResponse.bidId || bidResponse.requestId,
+    placementId: bidResponse.params
+      ? deepAccess(bidResponse, 'params.0.placementId')
+      : null,
+  };
+};
 
-function mapBidResponse(bidResponse, status) {
-  if (typeof bidResponse !== 'undefined') {
-    let bid = {
-      adUnitCode: bidResponse.adUnitCode,
-      gptSlotCode: getGptSlotInfoForAdUnitCode(bidResponse.adUnitCode).gptSlot || null,
-      auctionId: bidResponse.auctionId,
-      bidderCode: bidResponse.bidder,
-      cpm: bidResponse.cpm,
-      creativeId: bidResponse.creativeId,
-      currency: bidResponse.currency,
-      floorData: bidResponse.floorData,
-      mediaType: bidResponse.mediaType,
-      netRevenue: bidResponse.netRevenue,
-      requestTimestamp: bidResponse.requestTimestamp,
-      responseTimestamp: bidResponse.responseTimestamp,
-      status: bidResponse.status,
-      statusMessage: bidResponse.statusMessage,
-      timeToRespond: bidResponse.timeToRespond,
-      transactionId: bidResponse.transactionId
-    };
-    if (status !== 'bidwon') {
-      Object.assign(bid, {
-        bidId: status === 'timeout' ? bidResponse.bidId : bidResponse.requestId,
-        renderStatus: status === 'timeout' ? 3 : 2,
-        sizes: parseSizesInput(bidResponse.size).toString(),
+/**
+ * Track the events emitted by prebid and handle each case. See https://docs.prebid.org/dev-docs/publisher-api-reference/getEvents.html for more info
+ * @param {object} event the prebid event emmitted
+ * @param {string} event.eventType the type of the event
+ * @param {object} event.args the arguments of the emitted event
+ */
+const track = ({ eventType, args }) => {
+  switch (eventType) {
+    // handle invalid bids, and remove them from the adUnit cache
+    case CONSTANTS.EVENTS.BID_TIMEOUT:
+      args.map(extractBid).forEach((bid) => {
+        bid.renderStatus = 3;
+        auctionCache[bid.auctionId].bids.push(bid);
+      });
+      break;
+    // handle valid bid responses and record them as part of an auction
+    case CONSTANTS.EVENTS.BID_RESPONSE:
+      const bid = Object.assign(extractBid(args), { renderStatus: 2 });
+      auctionCache[bid.auctionId].bids.push(bid);
+      break;
+    // capture extra information from the auction, and if there were no bids
+    // (and so no chance of a win) send the auction
+    case CONSTANTS.EVENTS.AUCTION_END:
+      Object.assign(
+        auctionCache[args.auctionId].floorDetail,
```

Comment:
check for floorData exists from any of the bids

Reply:
done; we now check each ad unit for an appropriate bid payload and take the first configured floor data (so incorrect ad unit setups are avoided)